Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ccl/ui/text/CCLTextFormatter.kt

```diff
@@ -0,0 +1,194 @@
+            ?: return null
+
+        val text = pluralText(quantity, quantityText, locale)
+        return text
+            .replace("%@", "%s")
+            .format(*parameters.convertValues(locale))
+    }
+
+    private fun PluralText.quantity(): Int = quantity ?: quantityFromIndex()
+
+    private fun PluralText.quantityFromIndex(): Int {
+        val param = parameters[quantityParameterIndex ?: 0]
+        return when (param.type) {
+            Parameters.Type.NUMBER -> param.toNumber()
+            else -> error("`quantity` can't be derived from param=$param")
+        }
+    }
+
+    private suspend fun SystemTimeDependentText.formatSystemTimeDependent(): CCLText? =
+        runCatching {
+            val defaultParameters = getDefaultInputParameters(DateTime.now()).toObjectNode()
+            val allParameters = JsonNodeFactory.instance.objectNode()
+                .setAll<ObjectNode>(defaultParameters)
+                .setAll<ObjectNode>(parameters)
+            val output = cclJsonFunctions.evaluateFunction(functionName, allParameters)
+            mapper.treeToValue(output, CCLText::class.java)
+        }.onFailure {
+            Timber.e(it, "SystemTimeDependentText.formatSystemTimeDependent() failed.")
+        }.getOrNull()
+
+    private fun List<Parameters>.convertValues(locale: Locale): Array<Any> =
+        map { parameter -> parameter.covertValue(locale) }.toTypedArray()
+
+    private fun Parameters.covertValue(locale: Locale) = when (type) {
+        Parameters.Type.STRING -> value.toString()
+        Parameters.Type.NUMBER -> toNumber()
+        Parameters.Type.BOOLEAN -> toBoolean()
+        Parameters.Type.LOCAL_DATE -> toLocalDate(locale)
+        Parameters.Type.LOCAL_DATE_TIME -> toLocalDateTime(locale)
+        Parameters.Type.UTC_DATE -> toUTCDate(locale)
+        Parameters.Type.UTC_DATE_TIME -> toUTCDateTime(locale)
+    }
+
+    private fun Parameters.toUTCDateTime(locale: Locale): String {
+        return runCatching {
+            Instant.parse(value.toString()).run {
+                val date = toString(DateTimeFormat.shortDate().withLocale(locale))
+                val time = toString(DateTimeFormat.shortTime().withLocale(locale))
+                "$date, $time"
+            }
+        }.getOrElse {
+            Timber.e(it, "Parameters.toUTCDateTime() failed")
+            ""
+        }
+    }
+
+    private fun Parameters.toUTCDate(locale: Locale): String {
+        return runCatching {
+            Instant.parse(value.toString())
+                .toString(DateTimeFormat.shortDate().withLocale(locale))
+        }.getOrElse {
+            Timber.e(it, "Parameters.toUTCDate() failed")
+            ""
+        }
+    }
+
+    private fun Parameters.toLocalDateTime(locale: Locale): String {
+        return runCatching {
+            Instant.parse(value.toString()).toLocalDateTimeUserTz().run {
+                val date = toString(DateTimeFormat.shortDate().withLocale(locale))
```

Comment:
DateTime is a better choice if you are working with time zones

Reply:
done